pkg/syncer/cnsoperator/controller/csinodetopology/csinodetopology_controller.go

```diff
@@ -0,0 +1,336 @@
+		backOffDuration[instance.Name] = time.Second
+	}
+	timeout = backOffDuration[instance.Name]
+	backOffDurationMapMutex.Unlock()
+
+	// Fetch vsphere config
+	cfg, err := common.GetConfig(ctx)
+	if err != nil {
+		msg := "Failed to read CNS config"
+		log.Errorf("%s. Error: %v", msg, err)
+		_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+		return reconcile.Result{RequeueAfter: timeout}, nil
+	}
+	// Retrieve topology labels for nodeVM
+	if cfg.Labels.Zone == "" && cfg.Labels.Region == "" {
+		// Not a topology aware setup. No need to check for labels on nodeVM.
+		// Set the Status to Success and return
+		instance.Status.TopologyLabels = make([]csinodetopologyv1alpha1.TopologyLabel, 0)
+		err = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologySuccess, "")
+		if err != nil {
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+	} else if cfg.Labels.Zone != "" && cfg.Labels.Region != "" {
+		log.Infof("Detected a topology aware cluster")
+
+		// Fetch topology labels for nodeVM
+		topologyLabels, err := getNodeTopologyInfo(ctx, nodeVM, cfg)
+		if err != nil {
+			msg := fmt.Sprintf("failed to fetch topology information for the nodeVM with ID %q", nodeID)
+			log.Errorf("%s. Error: %v", msg, err)
+			_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Raise error if nodeVM does not have a topology label associated with each category in the
+		// vSphere config secret `Labels` section. For now, as we support only zone, region, hardcoded the value to 2
+		// TODO: Count the number of topology categories given and verify against that number.
+		if len(topologyLabels) != 2 {
+			msg := fmt.Sprintf("Detected a topology aware cluster. However, nodeVM with ID %q does not have a "+
+				"topology label for each category mentioned under the vSphere CSI config secret `Labels` section.", nodeID)
+			log.Error(msg)
+			_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+
+		// Update CSINodeTopology instance
+		instance.Status.TopologyLabels = topologyLabels
+		err = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologySuccess, "")
+		if err != nil {
+			return reconcile.Result{RequeueAfter: timeout}, nil
+		}
+	} else {
+		msg := "missing zone or region information in vSphere CSI config `Labels` section"
+		log.Error(msg)
+		_ = updateCRStatus(ctx, r, instance, csinodetopologyv1alpha1.CSINodeTopologyError, msg)
+		return reconcile.Result{}, errors.New(msg)
+	}
+
+	// On successful event, remove instance from backOffDuration
+	backOffDurationMapMutex.Lock()
+	delete(backOffDuration, instance.Name)
+	backOffDurationMapMutex.Unlock()
+	log.Infof("Successfully updated topology labels for nodeVM with ID %q", nodeID)
```

Comment:
How about calling `updateCRStatus` here with Success instead of lines 252 and 223

Reply:
The messages for the success calls are different. I will let them be in the `if..else` loop only.